base_layer/key_manager/src/cipher_seed.rs

```diff
@@ -155,223 +155,205 @@ impl CipherSeed {
+        encrypted_seed.push(CIPHER_SEED_VERSION);
+        encrypted_seed.append(&mut secret_data.to_vec());
+        encrypted_seed.append(&mut self.salt.to_vec());
 
         let mut crc_hasher = CrcHasher::new();
-        crc_hasher.update(final_seed.as_slice());
-        let checksum = crc_hasher.finalize();
-        final_seed.append(&mut checksum.to_le_bytes().to_vec());
-        Ok(final_seed)
+        crc_hasher.update(encrypted_seed.as_slice());
+        let mut checksum = crc_hasher.finalize().to_le_bytes().to_vec();
+        encrypted_seed.append(&mut checksum);
+
+        Ok(encrypted_seed)
     }
 
-    pub fn from_enciphered_bytes(enciphered_bytes: &[u8], passphrase: Option<String>) -> Result<Self, KeyManagerError> {
-        // 1 byte Version || 2 byte Birthday || 16 byte Entropy || 5 byte MAC || 5 byte salt || 4 byte CRC32
-        if enciphered_bytes.len() != 7 + CIPHER_SEED_ENTROPY_BYTES + CIPHER_SEED_SALT_BYTES + CIPHER_SEED_MAC_BYTES {
+    /// Recover a seed from encrypted data and a passphrase
+    pub fn from_enciphered_bytes(encrypted_seed: &[u8], passphrase: Option<String>) -> Result<Self, KeyManagerError> {
+        // Check the length: version, birthday, entropy, MAC, salt, checksum
+        if encrypted_seed.len() !=
+            1 + CIPHER_SEED_BIRTHDAY_BYTES +
+                CIPHER_SEED_ENTROPY_BYTES +
+                CIPHER_SEED_MAC_BYTES +
+                CIPHER_SEED_MAIN_SALT_BYTES +
+                CIPHER_SEED_CHECKSUM_BYTES
+        {
             return Err(KeyManagerError::InvalidData);
         }
 
-        if enciphered_bytes[0] != CIPHER_SEED_VERSION {
+        // We only support one version right now
+        let version = encrypted_seed[0];
+        if version != CIPHER_SEED_VERSION {
             return Err(KeyManagerError::VersionMismatch);
         }
 
-        let passphrase = passphrase.unwrap_or_else(|| DEFAULT_CIPHER_SEED_PASSPHRASE.to_string());
-
-        let mut body = enciphered_bytes.to_owned();
-        // extract 32 bit checksum
-        let checksum_vec = body.split_off(body.len() - 4);
+        let mut encrypted_seed = encrypted_seed.to_owned();
 
+        // Verify the checksum first, to detect obvious errors
+        let checksum = encrypted_seed.split_off(
+            1 + CIPHER_SEED_BIRTHDAY_BYTES +
+                CIPHER_SEED_ENTROPY_BYTES +
+                CIPHER_SEED_MAC_BYTES +
+                CIPHER_SEED_MAIN_SALT_BYTES,
+        );
         let mut crc_hasher = CrcHasher::new();
-        crc_hasher.update(body.as_slice());
-
-        let calculated_checksum = crc_hasher.finalize();
-
-        let mut checksum_bytes: [u8; 4] = [0u8; 4];
-        checksum_bytes.copy_from_slice(&checksum_vec[..4]);
-        let checksum = u32::from_le_bytes(checksum_bytes);
-
-        if calculated_checksum != checksum {
+        crc_hasher.update(encrypted_seed.as_slice());
+        let expected_checksum = crc_hasher.finalize().to_le_bytes().to_vec();
```

Comment:
nit: Allocation not necessary for byte comparison
```suggestion
        let expected_checksum = crc_hasher.finalize().to_le_bytes();
```